jaeger_client/sampler.py

```diff
@@ -421,56 +421,65 @@ def _sampling_request_callback(self, future):
-                self._parse_sampling_strategy(self.sampler, response.body)
+            sampling_strategies_response = json.loads(response.body)
         except Exception as e:
             self.error_reporter.error(
                 Metrics.SAMPLER_ERRORS, 1,
                 'Fail to parse sampling strategy '
                 'from jaeger-agent: %s [%s]', e, response.body)
             return
 
-        with self.lock:
-            if isinstance(sampler, AdaptiveSampler):
-                sampler.update(strategies)
-            elif self.sampler == sampler:
-                return
-            self.sampler = sampler
-        self.logger.debug('Tracing sampler set to %s', sampler)
-
-    def _poll_sampling_manager(self):
-        self.logger.debug('Requesting tracing sampler refresh')
-        fut = self._channel.request_sampling_strategy(
-            self.service_name, timeout=15)
-        fut.add_done_callback(self._sampling_request_callback)
+        self._update_sampler(sampling_strategies_response)
+        self.logger.debug('Tracing sampler set to %s', self.sampler)
 
-    def close(self):
+    def _update_sampler(self, response):
         with self.lock:
-            self.running = False
-            if self.periodic is not None:
-                self.periodic.stop()
+            try:
+                if response.get(OPERATION_SAMPLING_STR):
+                    self._update_adaptive_sampler(response.get(OPERATION_SAMPLING_STR))
+                else:
+                    self._update_rate_limiting_or_probabilistic_sampler(response)
+            except Exception as e:
+                self.error_reporter.error(
+                    Metrics.SAMPLER_ERRORS, 1,
+                    'Fail to update sampler'
+                    'from jaeger-agent: %s [%s]', e, response)
+
+    def _update_adaptive_sampler(self, per_operation_strategies):
+        if isinstance(self.sampler, AdaptiveSampler):
+            self.sampler.update(per_operation_strategies)
+        else:
+            self.sampler = AdaptiveSampler(per_operation_strategies, self.max_operations)
 
-    def _parse_sampling_strategy(self, sampler, body):
-        response = json.loads(body)
+    def _update_rate_limiting_or_probabilistic_sampler(self, response):
         s_type = response[STRATEGY_TYPE_STR]
-        operation_strategies = response.get(OPERATION_SAMPLING_STR)
-        if operation_strategies:
-            if isinstance(sampler, AdaptiveSampler):
-                return sampler, operation_strategies
-            return AdaptiveSampler(operation_strategies, self.max_operations), operation_strategies
         if s_type == SamplingManager.SamplingStrategyType.PROBABILISTIC:
             sampling_rate = response[PROBABILISTIC_SAMPLING_STR][SAMPLING_RATE_STR]
             if 0 <= sampling_rate <= 1.0:
-                return ProbabilisticSampler(rate=sampling_rate), None
+                self.sampler = ProbabilisticSampler(rate=sampling_rate)
```

Comment:
you can define `new_sampler` variable outside of `if` statement and do the `__eq__` test for all types. Otherwise you are forcing the assumption that replacing probabilistic sampling is safe - while it may be correct assumption, it's not necessary to force. Updating the sampler in one place in this function simplifies the logic.